front/src/applications/operationalStudies/components/MacroEditor/export.ts

```diff
@@ -0,0 +1,105 @@
+import {
+  osrdEditoastApi,
+  type Distribution,
+  type TrainScheduleBase,
+} from 'common/api/osrdEditoastApi';
+import type { AppDispatch } from 'store';
+
+import type { NGEDeleteEvent, NGEEvent, NGEUpdateEvent, TrainrunSectionOperation } from './types';
+
+const defautlPayloads = {
+  constraint_distribution: 'STANDARD' as Distribution,
+  rolling_stock_name: '',
+  start_time: '2024-07-15T08:00:00+02:00',
+};
+
+const ngeCreateOperation = (trainrunSection: TrainrunSectionOperation) => {
+  const { sourceNode, targetNode } = trainrunSection;
+  const [sourceTrigram, sourceSecondaryCode] = sourceNode.betriebspunktName.split('/');
+  const [targetTrigram, targetSecondaryCode] = targetNode.betriebspunktName.split('/');
+  const payloads: TrainScheduleBase[] = [
+    {
+      ...defautlPayloads,
+      path: [
+        {
+          trigram: sourceTrigram,
+          secondary_code: sourceSecondaryCode,
+          id: `${sourceNode.id}`,
+          deleted: false,
+        },
+        {
+          trigram: targetTrigram,
+          secondary_code: targetSecondaryCode,
+          id: `${targetNode.id}`,
+          deleted: false,
+        },
+      ],
+      train_name: trainrunSection.trainrun.name,
+    },
+  ];
+  return payloads;
+};
+
+const ngeUpdateOperation = (trainrunSections: TrainrunSectionOperation[]) => {
+  const path = trainrunSections.flatMap((section, index) => {
+    const [sourceTrigram, sourceSecondaryCode] = section.sourceNode.betriebspunktName.split('/');
+    const [targetTrigram, targetSecondaryCode] = section.targetNode.betriebspunktName.split('/');
+    const sourceNode = {
+      trigram: sourceTrigram,
+      secondary_code: sourceSecondaryCode,
+      id: `${section.sourceNode.id}-${index}`,
+    };
+    const targetNode = {
+      trigram: targetTrigram,
+      secondary_code: targetSecondaryCode,
+      id: `${section.targetNode.id}-${index}`,
+    };
+    if (index === trainrunSections.length - 1) {
+      return [sourceNode, targetNode];
+    }
+    return [sourceNode];
+  });
+  const [firstTrainrunSection] = trainrunSections;
+  const payloads: TrainScheduleBase = {
+    ...defautlPayloads,
+    path,
+    train_name: firstTrainrunSection.trainrun.name,
+  };
+  return payloads;
+};
+
+const ngeOperations = async (event: NGEEvent, dispatch: AppDispatch, timeTableId: number) => {
+  switch (event.type) {
+    case 'create': {
+      dispatch(
+        osrdEditoastApi.endpoints.postV2TimetableByIdTrainSchedule.initiate({
+          id: timeTableId!,
+          body: ngeCreateOperation(event.trainrunSection),
+        })
+      ).unwrap();
+      break;
+    }
+    case 'delete': {
+      const { trainrun } = event as NGEDeleteEvent;
+      await dispatch(
+        osrdEditoastApi.endpoints.deleteV2TrainSchedule.initiate({ body: { ids: [trainrun.id] } })
```

Comment:
Here we're assuming that `trainrun.id` is the OSRD train schedule ID. This works fine if the trainrun was imported from OSRD, but breaks when the trainrun was created inside NGE. Trainruns created inside NGE get an auto-incremented ID generated by NGE which don't match OSRD IDs.

(Same applies to `update`.)

I don't think we can easily fix this here, but it's potentially dangerous since we may delete completely unrelated trainruns (what NGE thinks trainrun 42 is may be a completely unrelated trainrun for OSRD).

(My plan to fix this completely would be to add a map of NGE <-> OSRD IDs.)

Reply:
An, ahem, quick'n'dirty way to ~~address~~ hack this would be to use a different type to host OSRD IDs, that way we can tell them apart from NGE IDs and bail out if we encounter one. Typically convert the OSRD ID to a string at import time, check if the ID is a string here, and convert it back to a number before passing it to the backend.

Please don't credit me for this if you end up implementing it.

Reply:
resolved by creating a map that gets the train schedule instead